### PR TITLE
Add substrings search path for constant like pattern

### PR DESCRIPTION
**Pull request by skadilover**
We noticed that some like constant-patterns can be equivalent to substrings search such as tpch queries Q13 and Q16: "%special%requests%" "%Customer%Complaints%"

1.  Add substring-search fast path for like with search pattern : "((%+[^%#_\\]+)+%+)"
2. Some OLAP engine supports "\_" for substring-search. In order to obtain the best performance, "\_" is not supported here, '#' is not support for the same reason too.
3. This PR will also affect single substring-search like tpch Q9,  if a fast-simd-strstr to speedup string-search will be added as follow-up, changes nothing right now.

benchmark result :
before:

tpchQuery9                                                 23.96ms     41.73
tpchQuery13                                               156.02ms      6.41
tpchQuery16Part                                             9.02ms    110.83
tpchQuery16Supplier                                       192.13ms      5.20

after:

tpchQuery9                                                 24.30ms     41.14
tpchQuery13                                                55.03ms     18.17
tpchQuery16Part                                             8.92ms    112.07
tpchQuery16Supplier                                        17.64ms     56.69

### CODE REVIEW

**Comment by skadilover** [comment]
@mbasmanova  @Yuhta 

Could you help to have a look at this PR ? The code format is still being fixed, I am not sure is it reasonable to add a fast-path here.
And could you help to look at the "load_aligned" method use in smid_strstr_memcmp is safe as it may load some unkown-memory address here.

File: velox/common/base/SimdUtil-inl.h [hunk 1]
```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
   return true;
 }
 
+namespace detail {
+template <typename T>
+T clearLeftmostSet(const T value) {
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1441]: These are well known patterns, just use them directly without helper functions.  Giving them names is rather confusing (first vs last vs left vs right)

File: velox/common/base/SimdUtil-inl.h [hunk 2]
```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+    const char* s,
+    size_t n,
+    const char* needle,
+    MEMCMP memcmp_fun) {
+  assert(k > 0);
+  assert(n > 0);
+
+  auto first = CharVector::broadcast(needle[0]);
+  auto last = CharVector::broadcast(needle[k - 1]);
+  for (size_t i = 0; i < n; i += CharVector::size) {
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1511]: `i + CharVector::size <= n` otherwise it's overreading, you need to handle trailing characters separately

File: velox/common/base/SimdUtil-inl.h [hunk 3]
```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+
+    case 10:
+      result = detail::smidStrstrMemcmp<10>(s, n, needle, memcmp8);
+      break;
+
+    case 11:
+      result = detail::smidStrstrMemcmp<11>(s, n, needle, memcmp9);
+      break;
+
+    case 12:
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1602]: You should extend it to 18, since 16 bytes fixed width comparison can be done on SSE in one instruction

File: velox/common/base/SimdUtil-inl.h [hunk 4]
```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+#define SIMD_STRSTR
+using CharVector = xsimd::batch<uint8_t, xsimd::avx2>;
+#elif XSIMD_WITH_NEON
+#define SIMD_STRSTR
+using CharVector = xsimd::batch<uint8_t, xsimd::neon>;
+#endif
+
+#ifdef SIMD_STRSTR
+size_t FOLLY_ALWAYS_INLINE
+smidStrstrAnysize(const char* s, size_t n, const char* needle, size_t k) {
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1472]: This can be combined with `smidStrstrMemcmp`.  Just pass `k` as both template parameter and runtime parameter.

**Comment by skadilover** [comment]
@Yuhta  update code , could you have a look at this once more ?

> i + CharVector::size <= n otherwise it's overreading, you need to handle trailing characters separately

I think this is the most interesting thing in this PR:
1. I add pageSafe check here now , like  ClickHouse does : https://github.com/ClickHouse/ClickHouse/blob/a51f867ce014a4cb8f97c46e004b90f5feafd80f/src/Common/StringSearcher.h#L31
```
    bool isPageSafe(const void * const ptr) const
    {
        return ((page_size - 1) & reinterpret_cast<std::uintptr_t>(ptr)) <= page_size - N;
    }
```
By this way,  load_unaligned is safe when overreading for they are in same memory page.

2. For StringSearch, ClickHouse use Volnitsky  https://github.com/ClickHouse/ClickHouse/blob/1513eda3efe3a24b7a87b9f0e7803bb8693d1bc8/src/Common/Volnitsky.h#L373 

Howevever, performance seems not best, Doris delete Volnitsky when they copy the StringSearch implement : https://github.com/apache/doris/blob/4e326e5f496109edb253d094869d057573ab7a59/be/src/vec/common/string_searcher.h#L44

And use first-two char compare, however, I test it , about 2x slow than this PR`s implement.

File: velox/common/base/SimdUtil-inl.h (hunk 3 above)
Comment (skadilover) [reply velox/common/base/SimdUtil-inl.h:1569]: Still need to do this , If I removed FixedMemCmp for they affect performance a little. 

For memcpy , also affect a little, however they are cheap to add, and I not remove it.

Comment (skadilover) [reply velox/common/base/SimdUtil-inl.h:1569]: tpchQuery2                                                 10.35ms     96.60
tpchQuery9                                                 16.38ms     61.04
tpchQuery13                                                28.56ms     35.01
tpchQuery14                                                 9.28ms    107.70
tpchQuery16Part                                             8.29ms    120.58
tpchQuery16Supplier                                        25.18ms     39.71
tpchQuery20                                                 7.31ms    136.84

**Comment by skadilover** [comment]
@Yuhta Yu  ping ～

**Comment by skadilover** [comment]
@Yuhta  ping here~ This PR is relative smaller and cheaper,  however, affect tpch-Q13`s performance a lot, about 2.5% up for total TPCH test.

**Comment by mbasmanova** [comment]
@xumingming James, would you take a first pass?

**Comment by mbasmanova** [comment]
@skadilover Are there 2 separate optimizations here? Can SIMD changes be extracted into a separate PR?

LIKE is a widely used construct. We need to make sure the change is 100% correct. Would you describe the testing approach and why do you believe it guarantees 100% correctness?

CC: @kagamiori @kgpai @amitkdutta

**Comment by mbasmanova** [comment]
@skadilover Please, make sure to document the new optimizations in https://facebookincubator.github.io/velox/functions/presto/regexp.html

File: velox/common/base/SimdUtil.h [hunk 5]
```diff
@@ -496,7 +496,8 @@ xsimd::batch<T, A> reinterpretBatch(xsimd::batch<U, A>, const A& = {});
 // equal. May address up to SIMD width -1 past end of either 'x' or 'y'.
 template <typename A = xsimd::default_arch>
 inline bool memEqualUnsafe(const void* x, const void* y, int32_t size);
-
+FOLLY_ALWAYS_INLINE size_t
```
Comment (mbasmanova) [inline velox/common/base/SimdUtil.h:499]: Add empty line and document this function. Make sure to spell out parameter names. Do not use single-letter names. Remove 'simd' prefix from the function name. This function is in velox::simd namespace, hence, "simd" is implied.

File: velox/common/base/tests/SimdUtilTest.cpp [hunk 6]
```diff
@@ -491,4 +491,25 @@ TEST_F(SimdUtilTest, memcpyTime) {
   LOG(INFO) << "simd=" << simd << " sys=" << sys;
 }
 
+TEST_F(SimdUtilTest, testSimdStrStr) {
```
Comment (mbasmanova) [inline velox/common/base/tests/SimdUtilTest.cpp:494]: Remove 'test' prefix and 'simd'. Both are implied.

File: velox/common/base/SimdUtil-inl.h [hunk 7]
```diff
@@ -1436,4 +1436,142 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+}
+
+template <bool compiled, size_t compiledNeedleSize>
+size_t FOLLY_ALWAYS_INLINE smidStrstrMemcmp(
+    const char* s,
+    size_t n,
+    const char* needle,
+    size_t needleSize) {
+  static_assert(compiledNeedleSize >= 2);
+  VELOX_CHECK(needleSize >= 2);
```
Comment (mbasmanova) [inline velox/common/base/SimdUtil-inl.h:1460]: VELOX_CHECK_GE

File: velox/common/base/SimdUtil-inl.h [hunk 8]
```diff
@@ -1436,4 +1436,142 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+  VELOX_CHECK(n > 0);
```
Comment (mbasmanova) [inline velox/common/base/SimdUtil-inl.h:1461]: VELOX_CHECK_GT

File: velox/common/base/SimdUtil-inl.h [hunk 9]
```diff
@@ -1436,4 +1436,142 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+  auto first = CharVector::broadcast(needle[0]);
+  auto last = CharVector::broadcast(needle[needleSize - 1]);
+  size_t i = 0;
+  // Fast path for page-safe data.
```
Comment (mbasmanova) [inline velox/common/base/SimdUtil-inl.h:1465]: Is this implementation a copy-paste from somewhere? If so, let's clarify the source. Does the test cover all code paths? How can this be verified?

**Review by mbasmanova — COMMENTED** [review]
@skadilover Some comments for simd-ized version of strstr. Let's extract this into a separate PR and a standalone benchmark.

File: velox/functions/lib/Re2Functions.h [hunk 10]
```diff
@@ -50,6 +49,8 @@ enum class PatternKind {
   kRelaxedSuffix,
   /// Patterns matching '%{c0}%', such as '%foo%%', '%%%hello%'.
   kSubstring,
+  /// Patterns matching '%{c0}%{c1}%', such as '%%foo%%bar%%', '%foo%bar%'.
+  kSubstrings,
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.h:53]: Why do we need both kSubstring and kSubstrings ? Can we combine these two?

File: velox/functions/lib/Re2Functions.cpp [hunk 11]
```diff
@@ -1741,22 +1769,47 @@ PatternMetadata PatternMetadata::relaxedSuffix(
...
+}
+
+std::vector<std::string> PatternMetadata::parseSubstrings(
+    const std::string_view& pattern) {
+  // Not support substrings-search with '_' for best performance.
+  static const re2::RE2 fullPattern(R"((%+[^%_#\\]+)+%+)");
+  static const re2::RE2 subPattern(R"((?:%+)([^%_#\\]+))");
+  re2::StringPiece full(pattern);
+  re2::StringPiece cur;
+  std::vector<std::string> substrings;
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.cpp:1792]: Wondering is there is a perf regression for the case where this pattern is not constant. Also, how do we handle custom escape characters? Is there some code that disables this optimization for that case?

File: velox/functions/lib/Re2Functions.cpp [hunk 12]
```diff
@@ -1973,6 +2026,14 @@ PatternMetadata determinePatternKind(
     return PatternMetadata::fixed("");
   }
 
+  // Fast path for substrings search.
+  {
+    auto substrings = PatternMetadata::parseSubstrings(pattern);
+    if (substrings.size() > 0) {
+      return PatternMetadata::substrings(substrings);
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.cpp:2033]: std::move to avoid copying strings

File: velox/functions/lib/tests/Re2FunctionsTest.cpp [hunk 13]
```diff
@@ -1500,5 +1510,22 @@ TEST_F(Re2FunctionsTest, split) {
   assertEqualVectors(expected, result);
 }
 
+TEST_F(Re2FunctionsTest, testParseSubstrings) {
```
Comment (mbasmanova) [inline velox/functions/lib/tests/Re2FunctionsTest.cpp:1513]: drop 'test'; please, review coding style guidelines and make sure the code is compliant

File: velox/functions/lib/tests/Re2FunctionsTest.cpp [hunk 14]
```diff
@@ -1500,5 +1510,22 @@ TEST_F(Re2FunctionsTest, split) {
...
+  auto test = [&](const std::string& input,
+                  const std::vector<std::string>& expected) {
+    ASSERT_EQ(PatternMetadata::parseSubstrings(input), expected);
+  };
+  // Cases that not supported by substrings-search.
+  test("%%", {});
+  // Not supports prefix.
+  test("aa%bb%%", {});
+  // Not supports sufix.
+  test("%aa%bb", {});
```
Comment (mbasmanova) [inline velox/functions/lib/tests/Re2FunctionsTest.cpp:1523]: Are these tests sufficient? How can this be verified? Should we test _ and custom escape chars?

File: velox/functions/lib/Re2Functions.cpp (hunk 11 above)
Comment (skadilover) [reply velox/functions/lib/Re2Functions.cpp:1792]: > Wondering is there is a perf regression for the case where this pattern is not constant. Also, how do we handle custom escape characters? Is there some code that disables this optimization for that case?

yeah,  so the real problem is that determinePatternKind will cause  perf regression, I think it could not be called for not constant, why we add it before,  and I think it should be removed for no-constant case

File: velox/functions/lib/Re2Functions.h (hunk 10 above)
Comment (skadilover) [reply velox/functions/lib/Re2Functions.h:53]: just for respect old code.

Comment (mbasmanova) [reply velox/functions/lib/Re2Functions.h:53]: Let's combine these options.

File: velox/functions/lib/Re2Functions.cpp (hunk 11 above)
Comment (mbasmanova) [reply velox/functions/lib/Re2Functions.cpp:1792]: > I think the parse code is not more complex than the left code in determinePatternKind

I wonder if that's indeed the case. I'm seeing that new code is using 2 regexes to match the pattern, applies second pattern multiple times, builds up a std::vector of strings using push_back w/o reserve + copies the strings. It might be worth benchmarking to determine whether this would cause significant regression and if so take care of that.

File: velox/common/base/SimdUtil-inl.h [hunk 15]
```diff
@@ -1436,4 +1436,142 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+  switch (k) {
+    case 0:
+      return 0;
+
+    case 1: {
+      const char* res = strchr(s, needle[0]);
+
+      return (res != nullptr) ? res - s : std::string::npos;
+    }
+#define FIXED_MEM_STRSTR(size)                                         \
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1542]: `VELOX_SIMD_STRSTR_CASE`

File: velox/common/base/SimdUtil-inl.h [hunk 16]
```diff
@@ -1436,4 +1436,142 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
...
+      FIXED_MEM_STRSTR(11)
+      FIXED_MEM_STRSTR(12)
+      FIXED_MEM_STRSTR(13)
+      FIXED_MEM_STRSTR(14)
+      FIXED_MEM_STRSTR(15)
+      FIXED_MEM_STRSTR(16)
+      FIXED_MEM_STRSTR(17)
+      FIXED_MEM_STRSTR(18)
+    default:
+      result = detail::smidStrstrMemcmp<false, 2>(s, n, needle, k);
```
Comment (Yuhta) [inline velox/common/base/SimdUtil-inl.h:1564]: aside: Wondering what is the minimal `k` to make KMP faster than this.  This is also related to content of the strings though.

File: velox/common/base/tests/SimdUtilTest.cpp (hunk 6 above)
Comment (Yuhta) [inline velox/common/base/tests/SimdUtilTest.cpp:494]: Since this is an almost drop-in replacement for `strstr`, can you add another test to generate a few thousands of random string pairs with various length, and check our result is the same as `strstr`?  Also make sure you cover corner cases like larger needle than stack, etc.

Comment (skadilover) [reply velox/common/base/tests/SimdUtilTest.cpp]: @Yuhta 
yeah,  I will add another pr later for simd_strstr later.

File: velox/functions/lib/tests/Re2FunctionsTest.cpp (hunk 14 above)
Comment (skadilover) [reply velox/functions/lib/tests/Re2FunctionsTest.cpp:1524]: > 'custom escape chars'
I think the case is already bypassed :
```
    BaseVector* escape = inputArgs[2].constantValue.get();
    if (!escape) {
      return std::make_shared<LikeGeneric>();
    }
```

File: velox/functions/lib/Re2Functions.cpp (hunk 11 above)
Comment (skadilover) [reply velox/functions/lib/Re2Functions.cpp:1788]: I assume that for non-constant-pattern  there is perf regression, so only add it for constant case.

File: velox/functions/lib/Re2Functions.h (hunk 10 above)
Comment (skadilover) [reply velox/functions/lib/Re2Functions.h:55]: Update : kSubstring is for non-constant case now.

I will add benchmark to measure determinePatternKind(), and disscuss it again,  I simple assume that parseSubstrings will add perf regression and only add it for constant-case.

**Review by skadilover — COMMENTED** [review]
@mbasmanova 
Update code:
1. remove simd_strstr.
2. Only add parseSubstrings for constant-case.

**Comment by skadilover** [comment]
@mbasmanova  Do you forget this PR?

**Comment by mbasmanova** [comment]
> I will add benchmark to measure determinePatternKind(), and disscuss it again, I simplely assume that parseSubstrings will add perf regression and only add it for constant-case.

@skadilover What did you find?

File: velox/functions/lib/Re2Functions.cpp [hunk 17]
```diff
@@ -650,6 +651,23 @@ bool matchSubstringPattern(
       std::string::npos);
 }
 
+bool matchSubstringsPattern(
+    const StringView& input,
+    const std::vector<std::string>& patterns) {
+  const char* data = input.data();
+  for (int i = 0; i < patterns.size(); i++) {
+    // A fast simd_strstr will add as following.
```
Comment (mbasmanova) [inline velox/functions/lib/Re2Functions.cpp:659]: What does this comment mean? Is it needed?

**Comment by skadilover** [comment]
@mbasmanova  update codes:
1.  update like documentation. 
2.  fix comments for kSubstrings as you mentioned.
3.  Regarding the problem of ‘determinePatternKind()’, let’s ignore it for now. I will study it clearly to see if it has performance problems in non-constant situations after I finish handling some existing work.

>I will add benchmark to measure determinePatternKind(), and disscuss it again, I simplely assume that parseSubstrings will add perf regression and only add it for constant-case.

>@skadilover What did you find?

File: velox/docs/functions/presto/regexp.rst [hunk 18]
```diff
@@ -32,6 +32,12 @@ limited to 20 different expressions per instance and thread of execution.
         SELECT like('abc', '%b%'); -- true
         SELECT like('a_c', '%#_%', '#'); -- true
 
+    String sequence search: There are some patterns that are equivalent to simple
```
Comment (mbasmanova) [inline velox/docs/functions/presto/regexp.rst:35]: This paragraph doesn't fit with the existing "Patterns 'hello',..." docs. It would be nice to make the whole doc consistent.

File: velox/functions/lib/Re2Functions.cpp [hunk 19]
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 #include "velox/functions/lib/Re2Functions.h"
+#include "velox/common/base/SimdUtil.h"
```
[the remark this answers is not in the document]
Comment (skadilover) [reply velox/functions/lib/Re2Functions.cpp]: er...., forget this after removed simdstrstr

File: velox/docs/functions/presto/regexp.rst (hunk 18 above)
Comment (skadilover) [reply velox/docs/functions/presto/regexp.rst:35]: updated with "hello velox"

**Comment by skadilover** [comment]
@mbasmanova  ping ～

File: velox/docs/functions/presto/regexp.rst (hunk 18 above)
Comment (mbasmanova) [inline velox/docs/functions/presto/regexp.rst:35]: It would be better to embed the new information into the existing flow.

> Not all patterns require compilation of regular expressions. Patterns ....

The new pattern needs to added to this list. The separate paragraph then can be about the limitations that apply to these special patterns. I.e. We would clarify separately whether special patterns allow for custom character or not.

Also, please, generate the docs to make sure they look nice.

Comment (skadilover) [reply velox/docs/functions/presto/regexp.rst]: image

I am not sure if I understand your meaning, add a list for all fast path in velox.

@mbasmanova  Could you look at this again ?

File: velox/docs/functions/presto/regexp.rst [hunk 20]
```diff
@@ -32,6 +32,22 @@ limited to 20 different expressions per instance and thread of execution.
         SELECT like('abc', '%b%'); -- true
         SELECT like('a_c', '%#_%', '#'); -- true
 
+Not all patterns require compilation of regular expressions. For the patterns without custom escape:
```
Comment (mbasmanova) [inline velox/docs/functions/presto/regexp.rst:35]: There is existing text

```
Not all patterns require
    compilation of regular expressions. Patterns 'hello', 'hello%', '_hello__%',....
```

We need to update it to add the new pattern. Currently, you are adding new text as if existing text doesn't exist or doesn't talk about similar thing.

Current:

```
Patterns 'hello', 'hello%', '_hello__%',
    '%hello', '%__hello_', '%hello%', where
```

New (proposed):

```
Patterns 'hello', 'hello%', '_hello__%',
    '%hello', '%__hello_', '%hello%', '%hello%velox%' where
```

File: velox/docs/functions/presto/regexp.rst [hunk 21]
```diff
@@ -32,6 +32,22 @@ limited to 20 different expressions per instance and thread of execution.
...
+
+| patterns  | description  | fast search mothod |
```
Comment (mbasmanova) [inline velox/docs/functions/presto/regexp.rst:37]: This table might be a nice alternative to current doc, but we'll need to iterate on columns, formatting and user-friendly descriptions. I suggest to do that in a separate PR.

**Comment by skadilover** [comment]
@mbasmanova 
Got it, thanks for detail the comment, updated code, would you check it again?

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Comment by Yuhta** [comment]
Is the regression in benchmark caused by the change?  If so we probably want to merge this after #10858

**Comment by skadilover** [comment]
> Is the regression in benchmark caused by the change? If so we probably want to merge this after #10858

It`s strange because Q2 and suffix is not supported by this change.

**Comment by Yuhta** [comment]
@skadilover It's just flakiness, disappeared after rerun

**Comment by skadilover** [comment]
>  It's just flakiness, disappeared after rerun

I think it`s nothing to do with this change.

File: velox/functions/lib/Re2Functions.cpp [hunk 22]
```diff
@@ -2154,6 +2201,14 @@ std::shared_ptr<exec::VectorFunction> makeLike(
 
   PatternMetadata patternMetadata = PatternMetadata::generic();
   try {
+    // Fast path for substrings search.
+    auto substrings =
```
Comment (JkSelf) [inline velox/functions/lib/Re2Functions.cpp:2205]: After picking this PR, Gluten will be a unit test that fails.

`Incorrect evaluation: ///__ LIKE %//%/% ESCAPE '/', actual: true, expected: false `

It seems that this implementation did not take into account the ESCAPE. In this example, the last slash (/) is the ESCAPE, which causes the last percent (%) to be a regular character, not a wildcard. Therefore, the result should be false, but Velox returns true.
@skadilover @mbasmanova @Yuhta @xiaoxmeng Do you have any input?

Comment (skadilover) [reply velox/functions/lib/Re2Functions.cpp:2205]: So sorry for the mistake, I mis-understand the code here: 
```
    BaseVector* escape = inputArgs[2].constantValue.get();
    if (!escape) {
      return std::make_shared<LikeGeneric>();
    }
```

and the ut just skiped by '#'
```
  testLike("a%c", "%#%%", '#', true);
  testLike("%cd", "%#%%", '#', true);
```

fix it here : https://github.com/facebookincubator/velox/pull/11091
